docs/implementation_design/plugin.md

```diff
@@ -1 +1,190 @@
-# Plugin
+# Plugin system
+
+GoShimmer is a complex application that is used in a research environment where requirements often changed and new ideas arise. 
+Plugin system allows to quickly and easily add and remove modules that need to be started. However, one thing that might be non-intuitive about the use of plugins is that it's taken to an extreme - everything is run through plugins. 
+The only code that is not executed through a plugin system is the code responsible for configuring and starting the plugins.
+All new future features added to the GoShimmer must be added by creating a new plugin. 
+
+
+## Plugin structure
+
+`Plugin` structure is defined as following.
+
+```go
+type Plugin struct {
+	Node   *Node
+	Name   string
+	Status int
+	Events pluginEvents
+	wg     *sync.WaitGroup
+}
+```
+
+Below is a brief description of each field: 
+* `Node` - contains a pointer to `Node` object which contains references to all the plugins and node-level logger. #TODO: figure out why it is there - not really used anywhere
+* `Name` - descriptive name of the plugin.
+* `Status` - flag indicating whether plugin is enabled or disabled.
+* `Events` - structure containing events used to properly deploy the plugin. Details described below.
+* `wg` - a private field containing WaitGroup. #TODO: figure out why it is there - not really used anywhere
+
+## Plugin events
+
+Each plugin defines 3 events: `Init`, `Configure`, `Run`. 
+Those events are triggered during different stages of node startup, but plugin doesn't have to define handlers for all of those events in order to do what it's been designed for.
+Execution order and purpose of each event is described below: 
+
+1. `Init` - is triggered almost immediately after a node is started. It's used in plugins that are critical for GoShimmer such as reading config file or initializing global logger. Most plugins don't need to use this event.
+2. `Configure` - this event is used to configure the plugin before it is started. It is used to define events related to internal plugin logic or initialize objects used by the plugin. 
+3. `Run` - this event is triggered as the last one. The event handler function contains main logic of the plugin. 
+   For many plugins, the event handler function creates a separate worker that works in the background, so that the handler function for one plugin can finish and allow other plugins be started.  
```

Comment:
```suggestion
   For many plugins, the event handler function creates a separate worker that works in the background, so that the handler function for one plugin can finish and allow other plugins to be started.  
```